src/main/webapp/app/exam/exam-scores/exam-scores.component.ts

```diff
@@ -737,4 +779,173 @@ export class ExamScoresComponent implements OnInit, OnDestroy {
         const percentage = this.noOfExamsFiltered && this.noOfExamsFiltered > 0 ? this.roundAndPerformLocalConversion((submissionCount * 100) / this.noOfExamsFiltered) : 0;
         return submissionCount + ' (' + percentage + '%)';
     }
+
+    /**
+     * fill ngxData with a default configuration. The assignment of the names is only a placeholder,
+     * they will be set to default labels in createChart.
+     * If a grading key exists, ngxData gets reset according to it in calculateFilterDependentStatistics.
+     * If no grading key exists, this default configuration is presented to the user.
+     * @private
+     */
+    private generateDefaultNgxChartsSetting(): void {
+        this.ngxData = [];
+        for (let i = 0; i < 100 / this.binWidth; i++) {
+            this.ngxData.push({ name: i.toString(), value: 0 });
+        }
+    }
+
+    /**
+     * Sets up the bar coloring
+     * If no grading scale exists, all bars representing a score < 40% are colored yellow in order to visualize that this is a critical performance
+     * If a grading scale exists that is bonus, all bars with a lower bound < 40% are colored yellow as well
+     * If a grading scale exists that is not bonus, all bars representing a score that does not pass the exam are colored red
+     * In either case, all bars above the thresholds remain grey
+     * @private
+     */
+    private setupChartColoring(): void {
+        this.ngxColor.domain = [];
+        if (!this.gradingScaleExists) {
+            for (let i = 0; i < 100 / this.binWidth; i++) {
+                if (i < 8) {
+                    this.ngxColor.domain.push(GraphColors.YELLOW);
+                } else {
+                    this.ngxColor.domain.push(GraphColors.GREY);
+                }
+            }
+        } else {
+            this.gradingScale!.gradeSteps.forEach((gradeStep) => {
+                let color;
+                if (this.isBonus) {
+                    if (gradeStep.gradeName === '0') {
+                        color = GraphColors.YELLOW;
+                    } else {
+                        color = GraphColors.GREY;
+                    }
+                } else {
+                    if (gradeStep.isPassingGrade) {
+                        color = GraphColors.GREY;
+                    } else {
+                        color = GraphColors.RED;
+                    }
+                }
```

Comment:
Option: This could maybe extracted into a separate method `getGradeStepColor` to reduce the nesting in this method.

Reply:
I agree that this method is rather hard to read currently. I extracted the color determination logic in an auxiliary method.